daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr.scala

```diff
@@ -126,30 +128,58 @@ object SExpr {
   /** Closure creation. Create a new closure object storing the free variables
     * in 'body'.
     */
-  final case class SEMakeClo(fv: Array[Int], arity: Int, body: SExpr)
+  final case class SEMakeClo(fvs: Array[SELoc], arity: Int, body: SExpr)
       extends SExpr
       with SomeArrayEquals {
 
     def execute(machine: Machine): Unit = {
-      def convertToSValues(fv: Array[Int], getEnv: Int => SValue) = {
-        val sValues = new Array[SValue](fv.length)
-        var i = 0
-        while (i < fv.length) {
-          sValues(i) = getEnv(fv(i))
-          i = i + 1
-        }
-        sValues
+      val sValues = new Array[SValue](fvs.length)
+      var i = 0
+      while (i < fvs.length) {
+        sValues(i) = fvs(i).lookup(machine)
+        i = i + 1
       }
-
-      val sValues = convertToSValues(fv, machine.getEnv)
       machine.returnValue = SPAP(PClosure(body, sValues), new util.ArrayList[SValue](), arity)
     }
   }
 
+  /** SELoc -- Reference to the runtime location of a variable.
+
+    This is the closure-converted form of SEVar. There are three sub-forms, with sufffix:
+    S/A/F, indicating [S]tack, [A]argument, or [F]ree variable captured by a closure.
+    */
+  sealed abstract class SELoc extends SExpr {
```

Comment:
I think we should remove this intermediate class and make the `SELocX` classes direct subclasses of `SExpr` with their own implementation of `execute`. This will make the code shorter and avoid one dynamic dispatch on each variable lookup. I would consider that a win-win situation. 😃

Reply:
Problem is that `SELoc` is also used here:
```
 final case class SEMakeClo(fvs: Array[SELoc], arity: Int, body: SExpr)
```
I could perhaps try hand inlining the calls to `lookup` from `execute` in each of the 3 sub classes. See if that has any effect...

Reply:
Nope.
In fact the time went up from 92 to 93 ms/op on the `CollectAuthority` bench.
But just noise I suspect.

Reply:
After last week... I have come to the opinion that trying to second guess the JVM on stuff like this is very unlikely to have noticeable results.